Subscribie/subscribie
# #805 #784 As a Subscriber I can settle a negative balance myself

Ref #805 

What's new?

- New Subscriber route http://127.0.0.1:5000/account/failed-invoices #805 
- New Subscriber route: http://127.0.0.1:5000/account/pay-invoice/in_<invoice-reference> #784 

## How to test

1. As a Subscriber, subscribe and [find and use test card which fails payments after inital card attachment](https://stripe.com/docs/testing).Take note of the failed invoice state `open` (As shop admin)

2. As a Subscriber, login to the [Subscriber login page](http://127.0.0.1:5000/account/login)
3. As a Subscriber, locate and visit the "Failed Invoices" page:
<img height="300px" src="https://user-images.githubusercontent.com/1718624/160305137-7e5eb7ca-94f1-467b-bfc1-407d7fdbdc32.png" />

5. As a Subscriber Click 'Pay now' on the failed invoices page
(failed invoices page)

<img height="300px" src="https://user-images.githubusercontent.com/1718624/160305210-8291ac5c-79b5-46fd-be37-4c85f871f9ff.png" />
6. Obvserve: You are redirected to Stripe invoice payment page
7. Use Stripe test card (e.g. the 4242 ... card) to complete the payment
8. Navigate back to shop (is this automatic?) , TODO: refresh invoices, observe the invoice is changed from status `open` to `paid`.

### Details

- New concept: "failed invoices"
- new database table `stripe_invoice` Purpose: To reduce round trip time fetching invoice information from Stripe each time (cache). This is a  fast cache (will *not* be up to date) until refreshed via a get via Stripe api. see [detail](https://github.com/Subscribie/subscribie/pull/811/files#diff-26a4fb09b7b90ccf42d6a13b35d380aebe44be88a5c06e7985bd988bc0349097)
  - This is a **cache** which must be refreshed (either manually or automatically) to fetch Stripe invoices (reglardless of if they are failed or not- all invoices are cached using [`get_stripe_invoices`](https://github.com/Subscribie/subscribie/pull/811/files#diff-6ef88f6680b4d380b20fe45a1890c33097b419042fd5ed45ac6c748432dff3f8R170)) into the table `stripe_invoice`
  - Failed Stripe invoices are fetched via [`get_stripe_failed_subscription_invoices`](https://github.com/Subscribie/subscribie/pull/811/files#diff-6ef88f6680b4d380b20fe45a1890c33097b419042fd5ed45ac6c748432dff3f8R252), which is fast because it uses the cached table `stripe_invoice` which stores stripe invoices. To determine a failed invoice (from Stripe's perspective is not simple: [see detail](https://github.com/Subscribie/subscribie/pull/811/files#diff-6ef88f6680b4d380b20fe45a1890c33097b419042fd5ed45ac6c748432dff3f8R252-R338))
- new database table `balance`
- new route guard [`@stripe_connect_id_required`](https://github.com/Subscribie/subscribie/pull/811/files#diff-69b7224a5125d1abc27988700d36aaba368cc5586891ae29239ce453b641cd68), purpose: see comment- does **note** require stripe connect is *activated*, only that a stripe connect id exists (and not needed for all pages)
- New admin route `/invoices/failed/` to show failed invoices, implemented by `get_stripe_failed_subscription_invoices`
